### PR TITLE
Regression models

### PR DESCRIPTION
**Pull request by amadeuspzs**
# Summary

**What/Why**

This PR includes the Regression models to baseline predicting Length of Stay (days) and an analysis of fairness.

**Your approach**

Two new notebooks and some updates to previous notebooks.

**What to review**

* 04 Feature engineering.ipynb - modified how sensitive features are stored for later use
* 05a Modelling - regression.ipynb
* 05b Modelling - regression - fairness.ipynb

# Checklists

This pull/merge request meets the following requirements:

- [X] code runs
- [X] [developments are ethical][data-ethics-framework]
- [X] developments are secure (no data or notebook output)
- [X] you have made proportionate checks that the code works correctly
- [X] [minimum usable documentation][agilemodeling] written in the `docs` folder

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: notebooks/05 Modelling - regression.ipynb [hunk 1]
```diff
@@ -0,0 +1,904 @@
...
+    "import matplotlib.pyplot as plt\n",
+    "import plotly.express as px\n",
+    "from sklearn.model_selection import train_test_split, GridSearchCV\n",
+    "from statsmodels.stats.outliers_influence import variance_inflation_factor\n",
+    "from sklearn import linear_model\n",
+    "from sklearn.dummy import DummyRegressor\n",
+    "from sklearn.ensemble import RandomForestRegressor\n",
+    "from sklearn.metrics import mean_absolute_error\n",
+    "from scipy.stats import kstest, shapiro, anderson\n",
+    "from catboost import CatBoostRegressor\n",
```
Comment (matthewcooper19) [inline notebooks/05 Modelling - regression.ipynb:42]: From what I can see, catboost is missing from the requirements.txt

File: notebooks/05b Modelling - regression fairness.ipynb [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "model = \"catboost\"\n",
+    "# Generate predictions\n",
+    "preds = np.clip(\n",
+    "    models[model][\"model\"].predict(X_test.drop(columns=sensitive_columns)), 0, None\n",
```
Comment (matthewcooper19) [inline notebooks/05b Modelling - regression fairness.ipynb:191]: Code falls over here, as trying to drop columns which have been one hot encoded

File: notebooks/05b Modelling - regression fairness.ipynb [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    "    data = pd.DataFrame(columns=[\"subfeature\", \"count\"])\n",
+    "    # iterate over all subcategories for the feature\n",
+    "    for subfeature in fairness_df[feature].unique():\n",
+    "        subset = fairness_df[fairness_df[feature] == subfeature]\n",
+    "        count = subset.LENGTH_OF_STAY.count()\n",
+    "        df = pd.DataFrame(data=[{\"subfeature\": subfeature, \"count\": count}])\n",
+    "        data = pd.concat([data, df])\n",
+    "\n",
+    "    fig = px.bar(data, x=\"subfeature\", y=\"count\", title=feature, orientation=\"v\")\n",
+    "    fig.add_hline(y=data[\"count\"].mean(), line_dash=\"dash\", line_color=\"green\")\n",
```
Comment (matthewcooper19) [inline notebooks/05b Modelling - regression fairness.ipynb:244]: Horizontal line only covers the first two bars in each of these plots, not sure if its supposed to appear like that

File: notebooks/05b Modelling - regression fairness.ipynb [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+    "for feature in sensitive_columns:\n",
+    "    data = pd.DataFrame(columns=[\"subfeature\", \"count\"])\n",
+    "    # iterate over all subcategories for the feature\n",
+    "    for subfeature in fairness_df[feature].unique():\n",
+    "        subset = fairness_df[fairness_df[feature] == subfeature]\n",
+    "        count = subset.LENGTH_OF_STAY.count()\n",
+    "        df = pd.DataFrame(data=[{\"subfeature\": subfeature, \"count\": count}])\n",
+    "        data = pd.concat([data, df])\n",
+    "\n",
+    "    fig = px.bar(data, x=\"subfeature\", y=\"count\", title=feature, orientation=\"v\")\n",
```
Comment (matthewcooper19) [inline notebooks/05b Modelling - regression fairness.ipynb:243]: The fix for vertical orientation looks good thanks!

File: notebooks/04 Feature engineering.ipynb [hunk 5]
```diff
@@ -78,7 +78,11 @@
...
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
+    "gather": {
+     "logged": 1652716832035
+    }
+   },
    "outputs": [],
    "source": [
     "derived_df = clean_data_df.copy()\n",
```
[the remark this answers is not in the document]
Comment (amadeuspzs) [reply notebooks/04 Feature engineering.ipynb:88]: Agreed. Actioned.

File: notebooks/05b Modelling - regression fairness.ipynb (hunk 3 above)
Comment (amadeuspzs) [reply notebooks/05b Modelling - regression fairness.ipynb:244]: Comment added ref: bug in Azure ML notebook rendering engine

File: notebooks/05a Modelling - regression.ipynb [hunk 6]
```diff
@@ -0,0 +1,1019 @@
...
+   "source": [
+    "# Feature names\n",
+    "coef = pd.DataFrame(data=list(X_train.columns))\n",
+    "# Feature importances, sorted\n",
+    "coef[\"coef\"] = models[model][\"model\"].feature_importances_\n",
+    "coef.sort_values(\"coef\", ascending=False, inplace=True)\n",
+    "coef.set_index(0, inplace=True)\n",
+    "# Plot interactive plot\n",
+    "fig = px.bar(coef, x=coef.index, y=\"coef\")\n",
+    "fig.show()"
```
[the remark this answers is not in the document]
Comment (amadeuspzs) [reply notebooks/05a Modelling - regression.ipynb:976]: Added comment on using mouseover

**Pull request merged**